### PR TITLE
Contracts & Harnesses for `add`, `addr`, and `align_offset`

### PR DESCRIPTION
**Pull request by QinyuanWu**
Towards https://github.com/model-checking/verify-rust-std/issues/53

## Changes
Three function contracts & four harnesses:
- added contract and harness for `non_null::add`
- added contract and harness for `non_null::addr`
- added contract and harnesses for `non_null::align_offset`, including both positive and negative harness that triggers panic. The ensures clause for `align_offset` is referenced from [`align_offset`](https://github.com/model-checking/verify-rust-std/pull/69/files) in `library/core/src/ptr/mod.rs`.

## Revalidation
To revalidate the verification results, run `kani verify-std -Z unstable-options "path/to/library"  -Z function-contracts -Z mem-predicates --harness ptr::non_null::verify`. This will run all six harnesses in the module. All default checks should pass:
```
SUMMARY:
 ** 0 of 1556 failed

VERIFICATION:- SUCCESSFUL
Verification Time: 0.28004378s

Complete - 6 successfully verified harnesses, 0 failures, 6 total.
```
### :exclamation: Warning
Running the above command with the default installed cargo kani will result in compilation error due to the latest merged from [PR#91](https://github.com/model-checking/verify-rust-std/pull/91). Detailed errors are commented under that PR. This issue is waiting to be resolved.

## TODO:

- Use `Layout` to create dynamically sized arrays in place of fixed size array in harnesses. This approach currently has errors and is documented in [discussion](https://github.com/model-checking/verify-rust-std/discussions/104).
- Verify multiple data types: these will be added in future PR.
- Add `requires` clause in contract to constrain `count` to be within object memory size: there is a current [issue](https://github.com/model-checking/verify-rust-std/discussions/99) with using `ub_checks::can_write` to get the object size. A workaround is implemented in the harness.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:282]: This requires clause is not needed since this is an invariant of this type.

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:283]: I think this can be strengthened to `result.get() == self.as_ptr() as usize`.

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:569]: This assert and the next one were added as part of our debugging of the failure in `offset` and should now be removed.

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:559]: As mentioned in https://github.com/model-checking/verify-rust-std/discussions/99#discussioncomment-10871535, the solution is to combine this requires clause with the `count.checked_mul` one to make sure they're properly applied.

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:1188]: This is not needed.

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:1827]: Is this used anywhere?

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:3]: This should not be needed.

**Review by zhassan-aws — COMMENTED** [review]
First round of comments.

File: library/core/src/ptr/non_null.rs
Comment (QinyuanWu) [reply library/core/src/ptr/non_null.rs:3]: WIthout `Self ` we get the error `error[E0433]: failed to resolve: use of undeclared crate or module `mem``

Comment (danielhumanmod) [inline library/core/src/ptr/non_null.rs:3]: Remove this import, we could use core::mem::* if I remember

File: library/core/src/ptr/non_null.rs
Comment (danielhumanmod) [inline library/core/src/ptr/non_null.rs:1190]: It seems a good reference, maybe we could introduce this to other team mates in the coding session

File: library/core/src/ptr/non_null.rs
Comment (QinyuanWu) [reply library/core/src/ptr/non_null.rs]: resolved with `same_allocation` and checking wrapping add.

File: library/core/src/ptr/non_null.rs
Comment (QinyuanWu) [reply library/core/src/ptr/non_null.rs]: removed

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:282]: The first conjunct (`result.get() != 0`) should not be needed as it is subsumed by the second conjunct (since `self.as_ptr()` is guaranteed to not be NULL).

Comment (zhassan-aws) [reply library/core/src/ptr/non_null.rs:282]: Also, is the cast to `*const ()` before casting to `usize` necessary?

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:561]: The second conjunct of the ensures clause (`kani::mem::same_allocation`) is not part of the function's spec.

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:1849]:
```suggestion
        let offset = kani::any_where(|x| *x <= SIZE as isize);
```

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:1850]: Please add this inside the `any_where` above.

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:1870]: It's more general to do:
```rust
        let x= kani::any::<usize>() as *mut i32;
```
and pass the result to `NonNull::new()`.

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:1879]: Would using:
```rust
        let raw_ptr = kani::any::<usize>() as *mut i32;
```
here as well work?

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:1892]: The comment incorrectly refers to an array.

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [inline library/core/src/ptr/non_null.rs:1896]: Same here.

File: library/core/src/ptr/non_null.rs
Comment (QinyuanWu) [reply library/core/src/ptr/non_null.rs:282]: @zhassan-aws Yes otherwise it gives the error `casting `*mut T` as `usize` is invalid`

Comment (QinyuanWu) [reply library/core/src/ptr/non_null.rs:282]: @zhassan-aws Yes otherwise it gives the error `casting `*mut T` as `usize` is invalid`.

File: library/core/src/ptr/non_null.rs
Comment (QinyuanWu) [reply library/core/src/ptr/non_null.rs:561]: The Safety requirement states: 
If the computed offset is non-zero, then `self` must be derived from a pointer to some [allocated object], and the entire memory range between `self` and the result must be in bounds of that allocated object. Doesn't that mean the resulted pointer should be pointing to the same object as the given self pointer?

File: library/core/src/ptr/non_null.rs
Comment (QinyuanWu) [reply library/core/src/ptr/non_null.rs:1870]: This works except we need to add `kani::assume(!x.is_null())` after this line so that we don't get failed check for calling `unwrap()` on a None object.

File: library/core/src/ptr/non_null.rs
Comment (celinval) [reply library/core/src/ptr/non_null.rs:282]: Can you use the `addr()` method instead?

File: library/core/src/ptr/non_null.rs
Comment (QinyuanWu) [reply library/core/src/ptr/non_null.rs:1892]: Thanks for catching!

File: library/core/src/ptr/non_null.rs
Comment (QinyuanWu) [reply library/core/src/ptr/non_null.rs:282]: @celinval the function under verification is `addr()`.

**Comment by QinyuanWu** [comment]
@zhassan-aws I tried to merge the 6 commits from the verify-std/main into this branch but when I run kani again I got the following error:
```
Kani Rust Verifier 0.56.0 (standalone)
error: cannot find a built-in macro with name `CoercePointee`
    --> /Users/owo/Desktop/verify-rust-std/library/core/src/marker.rs:1070:1
     |
1070 | / pub macro CoercePointee($item:item) {
1071 | |     /* compiler built-in */
1072 | | }
     | |_^

error: attributes starting with `rustc` are reserved for use by the `rustc` compiler
   --> /Users/owo/Desktop/verify-rust-std/library/core/src/alloc/layout.rs:562:32
    |
562 |     #[cfg_attr(not(bootstrap), rustc_const_stable_indirect)]
    |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: attributes starting with `rustc` are reserved for use by the `rustc` compiler
   --> /Users/owo/Desktop/verify-rust-std/library/core/src/alloc/layout.rs:485:32
    |
485 |     #[cfg_attr(not(bootstrap), rustc_const_stable_indirect)]
    |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^
......
```

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [reply library/core/src/ptr/non_null.rs:576]: Yes, but this is a _pre-condition_ of the function, and you already have it in the `requires` clause, so this should be sufficient.

File: library/core/src/ptr/non_null.rs
Comment (zhassan-aws) [reply library/core/src/ptr/non_null.rs]: Oh, yes. Adding the assumption is necessary.

**Comment by zhassan-aws** [comment]
You probably need to update the Kani version you're using. There's now a script that automates this process (https://github.com/model-checking/verify-rust-std/blob/main/scripts/run-kani.sh). If you run it, it should pull the correct Kani version.

**Comment by QinyuanWu** [comment]
> You probably need to update the Kani version you're using. There's now a script that automates this process (https://github.com/model-checking/verify-rust-std/blob/main/scripts/run-kani.sh). If you run it, it should pull the correct Kani version.
When I run the script I got the following error:
```
Using TOML file: tool_config/kani-version.toml
Using repository URL: https://github.com/model-checking/kani.git
Error: TOML file not found: tool_config/kani-version.toml
```

**Comment by zhassan-aws** [comment]
Did you run it from the root directory of the `verify-rust-std` repo? If not, can you try that?

**Comment by QinyuanWu** [comment]
> Did you run it from the root directory of the `verify-rust-std` repo? If not, can you try that?

@zhassan-aws The first time I ran I got
```
Error: Could not symlink include/cprover/api.h
Target /opt/homebrew/include/cprover/api.h
is a symlink belonging to cbmc@6.3.1. You can unlink it:
  brew unlink cbmc@6.3.1
```

So I entered the unlink command and re-ran the script but got:
```
ERROR: Could not find kani-driver binary.
Looked for: '/Users/owo/Desktop/verify-rust-std/kani_build/target/kani/bin/kani-driver'
Was Kani successfully built first?
```
My kani is built in ~/Desktop/kani.

**Comment by carolynzech** [comment]
> ```
> ERROR: Could not find kani-driver binary.
> Looked for: '/Users/owo/Desktop/verify-rust-std/kani_build/target/kani/bin/kani-driver'
> Was Kani successfully built first?
> ```
> 
> My kani is built in ~/Desktop/kani.

Try deleting the `kani_build` directory and re-running (see [these instructions](https://model-checking.github.io/verify-rust-std/tools/kani.html#step-2---run-the-kani-script-on-the-std-library)).

**Comment by QinyuanWu** [comment]
> > ```
> > ERROR: Could not find kani-driver binary.
> > Looked for: '/Users/owo/Desktop/verify-rust-std/kani_build/target/kani/bin/kani-driver'
> > Was Kani successfully built first?
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > My kani is built in ~/Desktop/kani.
> 
> @QinyuanWu Try deleting the `kani_build` directory and re-running (see [these instructions](https://model-checking.github.io/verify-rust-std/tools/kani.html#step-2---run-the-kani-script-on-the-std-library)).

@carolynzech  Debugged live with @zhassan-aws and resolved by manually re-building kani from the source. The run-kani script is having issue due to outdated cbmc-viewer installation that's still in the setup script of the kani repo and should be removed. Thanks!

**Comment by QinyuanWu** [comment]
Addressed previous PR comments and re-ran kani with successful verification:
```
SUMMARY:
 ** 0 of 141 failed

VERIFICATION:- SUCCESSFUL
Verification Time: 0.17998616s

Complete - 6 successfully verified harnesses, 0 failures, 6 total.
```

**Review by zhassan-aws — COMMENTED** [review]
Looks good!

It seems that the `stdarch` directory was accidentally updated. Can you revert the changes to it?

**Comment by QinyuanWu** [comment]
> Looks good!
> 
> It seems that the `stdarch` directory was accidentally updated. Can you revert the changes to it?

@zhassan-aws Done! Waiting for review from another team member.

**Review by zhassan-aws — APPROVED** [review]

File: library/core/src/ptr/non_null.rs
Comment (celinval) [inline library/core/src/ptr/non_null.rs:1217]: We should consider simplifying this later or at least move it to an auxiliary function.

File: library/core/src/ptr/non_null.rs
Comment (celinval) [inline library/core/src/ptr/non_null.rs:1906]: Instead of using `kani::assume()`, I would prefer to either implement Arbitrary or using:
```rust
let Some(nonnull) = NonNull::new(x) else { return; };
```

File: library/core/src/ptr/non_null.rs
Comment (celinval) [inline library/core/src/ptr/non_null.rs:1927]: You don't need this assumption

File: library/core/src/ptr/non_null.rs
Comment (celinval) [inline library/core/src/ptr/non_null.rs:1874]: Can you simplify this by using a non-deterministic usize and using wrapping_add instead?

**Review by celinval — APPROVED** [review]
Thank you. I added a few suggestions but not blockers

**Comment by QinyuanWu** [comment]
@zhassan-aws I just started to experience kani compiler error:
```
Kani Rust Verifier 0.56.0 (standalone)
error: attributes starting with `rustc` are reserved for use by the `rustc` compiler
    --> /Users/owo/Desktop/verify-rust-std/library/core/src/intrinsics.rs:2723:7
     |
2723 |     #[rustc_safe_intrinsic]
     |       ^^^^^^^^^^^^^^^^^^^^

error: attributes starting with `rustc` are reserved for use by the `rustc` compiler
    --> /Users/owo/Desktop/verify-rust-std/library/core/src/intrinsics.rs:2614:7
     |
2614 |     #[rustc_safe_intrinsic]
     |       ^^^^^^^^^^^^^^^^^^^^
```
Not sure if this is the same issue with PR #127

File: library/core/src/ptr/non_null.rs
Comment (QinyuanWu) [reply library/core/src/ptr/non_null.rs:1874]: @celinval This would be the most ideal case however this approach is essentially unbounded and will take too long for kani to execute.

**Comment by zhassan-aws** [comment]
> @zhassan-aws I just started to experience kani compiler error without any modification to the previous code:
> 
> ```
> Kani Rust Verifier 0.56.0 (standalone)
> error: attributes starting with `rustc` are reserved for use by the `rustc` compiler
>     --> /Users/owo/Desktop/verify-rust-std/library/core/src/intrinsics.rs:2723:7
>      |
> 2723 |     #[rustc_safe_intrinsic]
>      |       ^^^^^^^^^^^^^^^^^^^^
> 
> error: attributes starting with `rustc` are reserved for use by the `rustc` compiler
>     --> /Users/owo/Desktop/verify-rust-std/library/core/src/intrinsics.rs:2614:7
>      |
> 2614 |     #[rustc_safe_intrinsic]
>      |       ^^^^^^^^^^^^^^^^^^^^
> ```
> 
> Not sure if this is the same issue with PR #127

Can you try running using `./scripts/run_kani.sh`? We removed CBMC viewer from the installation steps, so it should hopefully work now.

File: library/core/src/ptr/non_null.rs
Comment (QinyuanWu) [reply library/core/src/ptr/non_null.rs]: Used `PointerGenerator` in the latest commit.

**Comment by QinyuanWu** [comment]
> > @zhassan-aws I just started to experience kani compiler error without any modification to the previous code:
> > ```
> > Kani Rust Verifier 0.56.0 (standalone)
> > error: attributes starting with `rustc` are reserved for use by the `rustc` compiler
> >     --> /Users/owo/Desktop/verify-rust-std/library/core/src/intrinsics.rs:2723:7
> >      |
> > 2723 |     #[rustc_safe_intrinsic]
> >      |       ^^^^^^^^^^^^^^^^^^^^
> > 
> > error: attributes starting with `rustc` are reserved for use by the `rustc` compiler
> >     --> /Users/owo/Desktop/verify-rust-std/library/core/src/intrinsics.rs:2614:7
> >      |
> > 2614 |     #[rustc_safe_intrinsic]
> >      |       ^^^^^^^^^^^^^^^^^^^^
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Not sure if this is the same issue with PR #127
> 
> Can you try running using `./scripts/run_kani.sh`? We removed CBMC viewer from the installation steps, so it should hopefully work now.

I resolved these errors by building kani from the features/verify-rust-std branch, but the errors in #127 still persists.

File: library/core/src/ptr/non_null.rs
Comment (carolynzech) [inline library/core/src/ptr/non_null.rs:10]:
```suggestion
use crate::{fmt, hash, intrinsics, ptr};
```

**Review by carolynzech — APPROVED** [review]
trigger approval workflow

**Pull request merged**
